### ISSUE TITLE
Kernel Lockdown is not enabled

### ISSUE THREAD

**Issue by erdnaxe** [NixOS/nixpkgs#251159]
### Describe the bug
When running `fwupdmgr security --force` on NixOS machines, I get a HSI runtime issue due to:
```
Linux kernel lockdown:         Failed
```
Looking at fwupdmgr doc (https://fwupd.github.io/libfwupdplugin/hsi.html#org.fwupd.hsi.Kernel.Lockdown), it looks like an interesting security mechanism introduced in Linux 5.4.

This problem has also been mentioned by @RaitoBezarius at https://nixos.paris/@raito/109671619222293620, but I see no corresponding opened issue in nixpkgs.

### Steps To Reproduce
1. Add `services.fwupd.enable = true;` to NixOS configuration,
2. Run `fwupdmgr security --force`,
3. Observe

### Expected behavior

Kernel lockdown should be available.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.45, NixOS, 23.05 (Stoat), 23.05.20230817.53baed0`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.5`
 - channels(root): `"nixos-23.05, nixos-unstable"`
 - channels(erdnaxe): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by RaitoBezarius** [comment]
Kernel lockdown cannot be enabled otherwise it will make the kernels unreproducible, I believe this is documented somewhere, but I cannot find the issue right now.

You would have to find the right trick to make the signing stuff irrelevant and still use kernel lockdown.

In the meantime, this will probably a lot of time to get it right as I don't have a high bandwidth.

**Comment by applePrincess** [comment]
> Kernel lockdown cannot be enabled otherwise it will make the kernels unreproducible, I believe this is documented somewhere, but I cannot find the issue right now.

Probably https://github.com/NixOS/nixpkgs/blob/febf8334aadeb281f1b76461be26d78d3c5914e6/pkgs/os-specific/linux/kernel/manual-config.nix#L126-L129

**Comment by Jackaed** [comment]
Note that if you want kernel lockdown, then you can enable it using a dummy kernel patch:

```nix
boot = {
   kernelPatches = [{
     name = "kernel-lockdown";
     patch = null;
     extraConfig = ''
       SECURITY_LOCKDOWN_LSM y
       MODULE_SIG y
     '';
   }];
};
```

I think it's really, really dumb that this has been disabled by default - particularly now that lanzaboote exists. Issue with using this dummy patch is that it means you have to compile the kernel every time there's a kernel update.

At the very least, it would be nice to have a build in nixpkgs that has this patch enabled (i.e. add an enableLockdown setting for all kernel builds) so it doesn't take 40 minutes for me to update every time there's a new version released.

**Comment by RaitoBezarius** [comment]
It's not really dumb, it's the problem of kernel signing and reproducibility of the kernel…

**Comment by Jackaed** [comment]
I don't think it's bad to have it disabled by default, but I don't see the harm in having an option to enable it without having to recompile the kernel locally. The question of whether it's possible to have a kernel with lockdown on nixos isn't really a question, you can already do it with the above patch. The only question is whether doing so forces you to locally recompile the kernel. I understand the issue it has with reproducibility.

**Comment by RaitoBezarius** [comment]
> I don't think it's bad to have it disabled by default, but I don't see the harm in having an option to enable it without having to recompile the kernel locally. The question of whether it's possible to have a kernel with lockdown on nixos isn't really a question, you can already do it with the above patch. The only question is whether doing so forces you to locally recompile the kernel. I understand the issue it has with reproducibility.

> At the very least, it would be nice to have a build in nixpkgs that has this patch enabled (i.e. add an enableLockdown setting for all kernel builds) so it doesn't take 40 minutes for me to update every time there's a new version released.

Yeah, but the cost of doing that is very expensive, we have to mind the resources taken to rebuild kernel every time there are new kernels for every kernel and the storage cost involved.

**Comment by Jackaed** [comment]
I do understand that - although that doesn't seem to have stopped some reasonably obscure alternate kernels from being packaged (liqorix kernel, xanmod kernel, zen kernel, bcachefs testing kernel, etc). Still, if this is too expensive to implement, then I understand it from that perspective, but I imagine that there would be more demand for people wanting kernel lockdown than, say, a very experimental new filesystem that almost no-one has heard of, let alone uses.

**Comment by RaitoBezarius** [comment]
You are asking all current kernels to exist in two versions, not adding *one* extra kernel, not the same?

>  but I imagine that there would be more demand for people wanting kernel lockdown than, say, a very experimental new filesystem that almost no-one has heard of, let alone uses.

I really find the way you bring things quite impolite to be honest. I am developing lanzaboote and bcachefs, sure you may not have heard about it, sure you may not be using it. But really, if you want kernel lockdown, send PRs to get reproducible kernel with kernel lockdown or contribute financially to the NixOS Foundation to ask more kernels in the cache.

If you cannot do either of those, please consider opening an Open Collective and crowdsource a project on making kernel lockdown better supported in nixpkgs out of the box. This is a volunteer project, run with volunteer efforts.

**Comment by Jackaed** [comment]
> You are asking all current kernels to exist in two versions, not adding one extra kernel, not the same?

Good point

> I really find the way you bring things quite impolite to be honest. I am developing lanzaboote and bcachefs

I apologize, I use lanzaboote daily, so thank you very much for the work you do on that, and I do think bcachefs is very promising - I shouldn't have criticized it like that. I recognize that this is a volunteer-run and struggling for funding, so I do sincerely apologize for being snarky in my prior comment. I doubt I'm experienced enough to attempt to work on reproducible kernels with lockdown enabled, but I will have a look at it.

**Comment by RaitoBezarius** [comment]
> > You are asking all current kernels to exist in two versions, not adding one extra kernel, not the same?
> 
> Good point
> 
> > I really find the way you bring things quite impolite to be honest. I am developing lanzaboote and bcachefs
> 
> I apologize, I use lanzaboote daily, so thank you very much for the work you do on that, and I do think bcachefs is very promising - I shouldn't have criticized it like that. I recognize that this is a volunteer-run and struggling for funding, so I do sincerely apologize for being snarky in my prior comment. I doubt I'm experienced enough to attempt to work on reproducible kernels with lockdown enabled, but I will have a look at it.

Thank you for your understanding!

**Comment by Lyndeno** [comment]
> Note that if you want kernel lockdown, then you can enable it using a dummy kernel patch:
> 
> ```nix
> boot = {
>    kernelPatches = [{
>      name = "kernel-lockdown";
>      patch = null;
>      extraConfig = ''
>        SECURITY_LOCKDOWN_LSM y
>        MODULE_SIG y
>      '';
>    }];
> };
> ```
> 
> I think it's really, really dumb that this can't be enabled without this patch - particularly now that lanzaboote exists. Issue with using this dummy patch is that it means you have to compile the kernel every time there's a kernel update.
> 
> At the very least, it would be nice to have a build in nixpkgs that has this patch enabled (i.e. add an enableLockdown setting for all kernel builds) so it doesn't take 40 minutes for me to update every time there's a new version released.

If one were to apply this patch. Is there another patch to enable Hibernation with lockdown? My understanding is lockdown disables hibernation.

Right now I am relying on nixpkgs disabling lockdown so my hibernation works. One day lockdown will work for reproducibility and will be enabled, and I will need a way to fix hibernation.

**Comment by Liassica** [comment]
@Lyndeno Simply compiling in the lockdown LSM (which is what this dummy patch does) won't break hibernation--you'd also need to explicitly enable lockdown on the kernel command line. If the lockdown LSM is ever added to the NixOS kernel, you wouldn't have to worry about it breaking hibernation.

**Comment by Lyndeno** [comment]
> @Lyndeno Simply compiling in the lockdown LSM (which is what this dummy patch does) won't break hibernation--you'd also need to explicitly enable lockdown on the kernel command line. If the lockdown LSM is ever added to the NixOS kernel, you wouldn't have to worry about it breaking hibernation.

Good to know, thanks. Reading the kernel mailing list about lockdown support gave me the impression that lockdown was enabled unconditionally when booting with secure boot.

But as long as it has to be explicitly enabled, or there is an option to manually disable it, then that is fine.

**Comment by Jackaed** [comment]
> If one were to apply this patch. Is there another patch to enable Hibernation with lockdown? My understanding is lockdown disables hibernation. 

Yes, there is another patch. I can't remember where I found it but here it is:

```patch
From e4db4f07e77feb1c126e7afbf441e9eae34b4e57 Mon Sep 17 00:00:00 2001
From: Kelvie Wong <kelvie@kelvie.ca>
Date: Sun, 2 Oct 2022 13:23:25 -0700
Subject: [PATCH] Add a lockdown_hibernate parameter

This allows the user to tell the kernel that they know better (namely,
they secured their swap properly), and that it can enable hibernation,
enabled the lockdown_hibernate kernel parameter.

Signed-off-by: Kelvie Wong <kelvie@kelvie.ca>
---
 Documentation/admin-guide/kernel-parameters.txt |  5 +++++
 kernel/power/hibernate.c                        | 10 +++++++++-
 2 files changed, 14 insertions(+), 1 deletion(-)

diff --git a/Documentation/admin-guide/kernel-parameters.txt b/Documentation/admin-guide/kernel-parameters.txt
index 426fa892d311..54785faba9e0 100644
--- a/Documentation/admin-guide/kernel-parameters.txt
+++ b/Documentation/admin-guide/kernel-parameters.txt
@@ -2804,6 +2804,11 @@
 			to extract confidential information from the kernel
 			are also disabled.
 
+	lockdown_hibernate	[HIBERNATION]
+			Enable hibernation even if lockdown is enabled. Enable this only if
+			your swap is encrypted and secured properly, as an attacker can
+			modify the kernel offline during hibernation.
+
 	locktorture.nreaders_stress= [KNL]
 			Set the number of locking read-acquisition kthreads.
 			Defaults to being automatically set based on the
diff --git a/kernel/power/hibernate.c b/kernel/power/hibernate.c
index 89c71fce225d..2221c531d54c 100644
--- a/kernel/power/hibernate.c
+++ b/kernel/power/hibernate.c
@@ -36,6 +36,7 @@
 #include "power.h"
 
 
+static int lockdown_hibernate;
 static int nocompress;
 static int noresume;
 static int nohibernate;
@@ -82,7 +83,7 @@ void hibernate_release(void)
 bool hibernation_available(void)
 {
 	return nohibernate == 0 &&
-		!security_locked_down(LOCKDOWN_HIBERNATION) &&
+		(lockdown_hibernate || !security_locked_down(LOCKDOWN_HIBERNATION)) &&
 		!secretmem_active() && !cxl_mem_active();
 }
 
@@ -1340,6 +1341,12 @@ static int __init nohibernate_setup(char *str)
 	return 1;
 }
 
+static int __init lockdown_hibernate_setup(char *str)
+{
+	lockdown_hibernate = 1;
+	return 1;
+}
+
 __setup("noresume", noresume_setup);
 __setup("resume_offset=", resume_offset_setup);
 __setup("resume=", resume_setup);
@@ -1347,3 +1354,4 @@ __setup("hibernate=", hibernate_setup);
 __setup("resumewait", resumewait_setup);
 __setup("resumedelay=", resumedelay_setup);
 __setup("nohibernate", nohibernate_setup);
+__setup("lockdown_hibernate", lockdown_hibernate_setup);
-- 
2.37.3
```

To use it, you also need to include `lockdown_hibernate=1` in your kernel launch params. I would only recommend doing this if you use an encrypted swap - otherwise it kind of defeats the purpose of kernel lockdown. There's a reason lockdown disables hibernation - but that reason doesn't apply if you're using an encrypted swap.

If you want to get around the fact that this would require a password when resuming from hibernation, I would setup decryption from the TPM.

**Comment by raboof** [comment]
There is some movement on combining lockdown and still having a reproducible kernel upstream, https://lore.kernel.org/all/20250120-module-hashes-v2-0-ba1184e27b7f@weissschuh.net/

I'm taking it for a spin with:

```
  boot.kernelPackages = pkgs.linuxPackagesFor (
    pkgs.linuxKernel.kernels.linux_testing.overrideAttrs(a: {
      # avoid invalidating module hashes
      # TODO check if we should make this more fine-grained to
      # to avoid undesirable side effects.
      dontStrip = true;

      postPatch = a.postPatch + ''
        chmod a+x scripts/module-hashes.sh
        patchShebangs scripts/module-hashes.sh
      '';
    }));
  boot.kernelPatches = [
    {
      patch = pkgs.fetchpatch {
        name = "module-hashes-v2-1";
        url = "https://lore.kernel.org/all/20250120-module-hashes-v2-1-ba1184e27b7f@weissschuh.net/raw";
        hash = "sha256-6BcMDbaoetRwv0GlksAgp1k638nDMo0Svn7oXklFbKs=";
      };
      extraConfig = ''
        MODULE_SIG n
        MODULE_HASHES y
        MODULE_SIG_POLICY y
        SECURITY_LOCKDOWN_LSM y
      '';
    }
    {
      patch = pkgs.fetchpatch {
        name = "module-hashes-v2-2";
        url = "https://lore.kernel.org/all/20250120-module-hashes-v2-2-ba1184e27b7f@weissschuh.net/raw";
        hash = "sha256-PQ1YsK+aqTI57qMPUh/MtLS97iU17MyjGLI2W4p3Y8I=";
      };
    }
      patch = pkgs.fetchpatch {
        name = "module-hashes-v2-3";
        url = "https://lore.kernel.org/all/20250120-module-hashes-v2-3-ba1184e27b7f@weissschuh.net/raw";
        hash = "sha256-qwcOU68tY9td9IllaP2kjeTz/xgGZSgSnyp2gOxxxGk=";
      };
    }
    {
      patch = pkgs.fetchpatch {
        name = "module-hashes-v2-4";
        url = "https://lore.kernel.org/all/20250120-module-hashes-v2-4-ba1184e27b7f@weissschuh.net/raw";
        hash = "sha256-RAwao/bRAA7tPKHS6DXKODdy9DHDvi0q/VRMr+8aD1E=";
      };
    }
    {
      patch = pkgs.fetchpatch {
        name = "module-hashes-v2-5";
        url = "https://lore.kernel.org/all/20250120-module-hashes-v2-5-ba1184e27b7f@weissschuh.net/raw";
        hash = "sha256-HlGkral6snZnxaNvjWR6S91Fh1jKdMj500LBNASmunc=";
      };
    }
    {
      patch = pkgs.fetchpatch {
        name = "module-hashes-v2-6";
        url = "https://lore.kernel.org/all/20250120-module-hashes-v2-6-ba1184e27b7f@weissschuh.net/raw";
        hash = "sha256-uzsBaC5CAa/hCi5o2ziRCaelmwvng1pRJEcPqQBQ/sc=";
      };
    }
  ];
```

This seems to work at first glance: indeed loading modules from my system kernel works, but loading a completely different on gives the slightly-confusing `modprobe: ERROR: could not insert 'btusb': Required key not available` (I checked, indeed it returns this error when the hash is not found ;) )
